skimage/data/_fetchers.py

```diff
@@ -152,21 +147,35 @@ def _skip_pytest_case_requiring_pooch(data_filename):
                     allow_module_level=True)
 
 
-def _fetch(data_filename):
+def _ensure_data_dir(*, target_dir):
+    """Prepare local cache directory if it doesn't exist already."""
+    os.makedirs(target_dir, exist_ok=True)
+    readme_src = osp.join(_DISTRIBUTION_DIR, "data/README.txt")
+    readme_dest = osp.join(target_dir, "README.txt")
+    if not osp.exists(readme_dest):
+        shutil.copy2(readme_src, readme_dest)
+
+
+def _fetch(data_filename, *, copy_legacy_to_cache=False):
```

Comment:
Can we get rid of this flag?

Reply:
Replying here because this touches on other questions you ask.

> But I suppose I am confused, because I thought the purpose of the PR would be to not create a cache directory or copy files into it until strictly necessary. I.e., if a package version is available, that should be served up instead of even touching the cache.

No you got that right. I'm trying to address two demands here:

1. Requesting a legacy dataset (the ones included in the distributed archives) from `_fetch` , serves a path without triggering the creation of the cache directory.
2. When `download_all(directory=...)` is called, put all files including legacy datasets into that directory. 

We can remove the flag if we update `download_all` so that 2. is still met. I'm happy to update the function accordingly. Following this PR, I wanted to ask if there is some interest in a larger refactoring anyway.

Reply:
That makes sense; yes, I think that would simplify things a bit, what do you think?

Reply:
Done.

Reply:
Note that I discovered two additional bugs in the old implementation of `download_all` while doing so: 

1. Running
      ```python
      import skimage as ski
      ski.data.download_all()
      ski.data.download_all(directory="example_dir")
      ```
      would not create anything in "example_dir" because `_fetch` would always
      return the cached entry before ever invoking pooches cache mechanism to
      place it in "example_dir". This is addressed by https://github.com/scikit-image/scikit-image/pull/6886/commits/0ddf46713dc9114be4532af70a64617a0d7c4dc8.

2. Running
    ```python
    import skimage as ski
    ski.data.download_all("~/skimage-data")
    ```
    will place files at two locations: files in the distribution are placed in "[working_dir]/~/skimage-data" while files downloaded with pooch are placed in  `/home/[user]/skimage-data`. I think this is because our old `os.path` machinery doesn't resolve `~` while pooch uses `pathlib` which does.

So, should we switch to `pathlib`? :pray: :sweat_smile:

Reply:
Haha, to keep this from blowing up, perhaps just `os.path.expanduser`.

Reply:
Good catches, btw!

Reply:
Okay :disappointed: , but honestly this is a bit of a mess. I am testing most of this stuff via console, debugging and checking actual folders. I would sincerely recommend to refactor this following this. Perhaps not with utmost priority but I wouldn't be surprised at all if there are more subtle edge cases that we didn't discover yet. :see_no_evil:

Reply:
Yes, that's a good idea: i.e., break it up into a utility function that handles pathnames (easily testable), and then some machinery for fetching from a fake cache perhaps.

If you want to do that refactor here, that'd be fine; we're not in a big rush.

Reply:
I'm happy to do the refactoring. But the CI is finally not complaining, so I wouldn't mind for this fix to go in in case the follow-up work get's stalled for some reason.